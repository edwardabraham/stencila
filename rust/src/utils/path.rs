use lexiclean::Lexiclean;
use std::path::{Path, PathBuf};

/// Merge paths
///
/// Differs from `path.join()` in that it joins the *parent* of `path1`, not `path1` itself.
/// Does not canonicalize the path, or check for its existence.
pub fn merge<P1: AsRef<Path>, P2: AsRef<Path>>(path1: P1, path2: P2) -> PathBuf {
    path1.as_ref().join("..").join(path2).lexiclean()
}

#[cfg(test)]
mod test {
    use super::*;
    use path_slash::PathBufExt;
    use pretty_assertions::assert_eq;

    #[test]
    fn test_merge() {
        assert_eq!(merge("a", "c").to_slash_lossy(), "c");
        assert_eq!(merge("a/b", "c").to_slash_lossy(), "a/c");
        assert_eq!(merge("a/b/../d/e", "c").to_slash_lossy(), "a/d/c")
    }
}
