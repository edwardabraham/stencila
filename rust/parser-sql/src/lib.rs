use std::path::Path;

use parser_treesitter::{
    common::{
        eyre::Result,
        once_cell::sync::Lazy,
        regex::{Captures, Regex},
        tracing,
    },
    formats::Format,
    graph_triples::{relations, resources, Resource, ResourceInfo},
    resource_info, Parser, ParserTrait, TreesitterParser,
};

/// A parser for SQL
pub struct SqlParser;

static LANG: Lazy<String> = Lazy::new(|| Format::SQL.spec().title);

const QUERY: &str = include_str!("query.scm");

static PARSER: Lazy<TreesitterParser> =
    Lazy::new(|| TreesitterParser::new(tree_sitter_sql::language(), QUERY));

impl ParserTrait for SqlParser {
    fn spec() -> Parser {
        Parser {
            language: LANG.clone(),
        }
    }

    fn parse(resource: Resource, path: &Path, code: &str) -> Result<ResourceInfo> {
        // Replace named bindings e.g. `$par_a` with numeric bindings e.g. `$1` so that
        // they are recognized by `tree-sitter-sql`.
        let mut bindings = Vec::new();
        static BINDING_REGEX: Lazy<Regex> = Lazy::new(|| {
            Regex::new(r"\$([a-zA-Z_][a-zA-Z_0-9]*)").expect("Unable to create regex")
        });
        let sql = BINDING_REGEX.replace_all(code, |captures: &Captures| {
            let name = captures[1].to_string();
            bindings.push(name);
            ["$", &bindings.len().to_string()].concat()
        });

        let code = sql.as_bytes();
        let tree = PARSER.parse(code);
        let matches = PARSER.query(code, &tree);

        let relations = matches
            .iter()
            .filter_map(|(pattern, captures)| {
                let relation = match pattern {
                    1 => relations::assigns(captures[0].range),
                    2 => relations::uses(captures[0].range),
                    3 => relations::alters(captures[0].range),
                    4 => relations::uses(captures[0].range),
                    _ => return None,
                };
                let name = match pattern {
                    4 => match captures[0].text[1..].parse::<usize>() {
                        Ok(index) => &bindings[index - 1],
                        Err(error) => {
                            tracing::error!(
                                "Unexpectedly unable to parse binding as integer index: {}",
                                error
                            );
                            return None;
                        }
                    },
                    _ => &captures[0].text,
                };
                let kind = match pattern {
                    4 => "",
                    _ => "Datatable",
                };
                Some((relation, resources::symbol(path, name, kind)))
            })
            .collect();

        let resource_info = resource_info(
            resource,
            path,
            &Self::spec().language,
            code,
            &tree,
            &["comment"],
            matches,
            0,
            relations,
        );
        Ok(resource_info)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use test_snaps::{insta::assert_json_snapshot, snapshot_fixtures};
    use test_utils::fixtures;

    #[test]
    fn parse_sql_fragments() {
        snapshot_fixtures("fragments/sql/*.sql", |path| {
            let code = std::fs::read_to_string(path).expect("Unable to read");
            let path = path.strip_prefix(fixtures()).expect("Unable to strip");
            let resource = resources::code(path, "", "SoftwareSourceCode", Some("SQL".to_string()));
            let resource_info = SqlParser::parse(resource, path, &code).expect("Unable to parse");
            assert_json_snapshot!(resource_info);
        })
    }
}
