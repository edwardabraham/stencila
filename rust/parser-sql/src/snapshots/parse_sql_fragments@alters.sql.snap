---
source: rust/parser-sql/src/lib.rs
assertion_line: 76
expression: resource_info
input_file: fixtures/fragments/sql/alters.sql
---
{
  "resource": {
    "type": "Code",
    "path": "fragments/sql/alters.sql",
    "id": "",
    "kind": "SoftwareSourceCode",
    "language": "SQL"
  },
  "relations": [
    [
      {
        "type": "Alters",
        "range": [
          0,
          12,
          0,
          20
        ]
      },
      {
        "type": "Symbol",
        "path": "fragments/sql/alters.sql",
        "name": "table_i1",
        "kind": "Datatable"
      }
    ],
    [
      {
        "type": "Alters",
        "range": [
          1,
          12,
          1,
          20
        ]
      },
      {
        "type": "Symbol",
        "path": "fragments/sql/alters.sql",
        "name": "table_i2",
        "kind": "Datatable"
      }
    ],
    [
      {
        "type": "Alters",
        "range": [
          2,
          12,
          2,
          20
        ]
      },
      {
        "type": "Symbol",
        "path": "fragments/sql/alters.sql",
        "name": "table_i3",
        "kind": "Datatable"
      }
    ],
    [
      {
        "type": "Uses",
        "range": [
          2,
          35,
          2,
          42
        ]
      },
      {
        "type": "Symbol",
        "path": "fragments/sql/alters.sql",
        "name": "table_a",
        "kind": "Datatable"
      }
    ],
    [
      {
        "type": "Alters",
        "range": [
          4,
          7,
          4,
          15
        ]
      },
      {
        "type": "Symbol",
        "path": "fragments/sql/alters.sql",
        "name": "table_u1",
        "kind": "Datatable"
      }
    ],
    [
      {
        "type": "Alters",
        "range": [
          6,
          12,
          6,
          20
        ]
      },
      {
        "type": "Symbol",
        "path": "fragments/sql/alters.sql",
        "name": "table_d1",
        "kind": "Datatable"
      }
    ],
    [
      {
        "type": "Alters",
        "range": [
          7,
          12,
          7,
          20
        ]
      },
      {
        "type": "Symbol",
        "path": "fragments/sql/alters.sql",
        "name": "table_d2",
        "kind": "Datatable"
      }
    ],
    [
      {
        "type": "Alters",
        "range": [
          9,
          11,
          9,
          19
        ]
      },
      {
        "type": "Symbol",
        "path": "fragments/sql/alters.sql",
        "name": "table_p1",
        "kind": "Datatable"
      }
    ]
  ],
  "compile_digest": "17352080989351478921.15361264440694782324.0.0.0",
  "tags": []
}
